### PR TITLE
Add a min version check to the resolver

### PR DESCRIPTION
**Pull request by livarcocc**
@dotnet/dotnet-cli @nguerrera 

**Customer scenario**

Planned feature work that will prevent an old Sdk from being used when the project declares that it requires a newer one.

**Bugs this fixes**

Fixes https://github.com/dotnet/cli/issues/6555

**Workarounds, if any**

N/A

**Risk**

Small. Might mean that VS won't find a CLI if we make the comparison check wrong. Added tests to cover this in the CLI.

**Performance impact**

no

**Root cause analysis**

N/A

**How was the bug found?**

Planned feature work.

### CODE REVIEW

**Comment by livarcocc** [comment]
@MattGertz for approval.

File: src/Microsoft.DotNet.MSBuildSdkResolver/MSBuildSdkResolver.cs [hunk 1]
```diff
@@ -53,14 +53,26 @@ public override SdkResult Resolve(SdkReference sdkReference, SdkResolverContext
...
 
+            if (!string.IsNullOrEmpty(sdkReference.MinimumVersion) && 
+                SemanticVersion.Parse(netcoreSdkVersion) < SemanticVersion.Parse(sdkReference.MinimumVersion))
+            {
+                return factory.IndicateFailure(
+                    new[]
+                    {
+                        $"Version {netcoreSdkVersion} of the SDK is smaller than the minimum version"
+                        + $" {sdkReference.MinimumVersion} requested. Check that a recent enough .NET Core SDK is"
+                        + " installed and/or increase the version specified in global.json."
```
Comment (nguerrera) [inline src/Microsoft.DotNet.MSBuildSdkResolver/MSBuildSdkResolver.cs]: " installed, increase the minimum version specified in the project, or increase the version specified in global.json"

File: src/Microsoft.DotNet.MSBuildSdkResolver/MSBuildSdkResolver.cs [hunk 2]
```diff
@@ -53,14 +53,26 @@ public override SdkResult Resolve(SdkReference sdkReference, SdkResolverContext
                 netcoreSdkVersion = new DirectoryInfo(netcoreSdkDir).Name;;
             }
 
+            if (!string.IsNullOrEmpty(sdkReference.MinimumVersion) && 
+                SemanticVersion.Parse(netcoreSdkVersion) < SemanticVersion.Parse(sdkReference.MinimumVersion))
```
Comment (nguerrera) [inline src/Microsoft.DotNet.MSBuildSdkResolver/MSBuildSdkResolver.cs]: You made SemanticVersion.Parse return null for invalid versions. Won't this null ref in that case?

File: src/Microsoft.DotNet.MSBuildSdkResolver/SemanticVersion.cs [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+        {
+            return Compare(s1, s2) >= 0;
+        }
+
+        public static bool operator <=(SemanticVersion s1, SemanticVersion s2)
+        {
+            return Compare(s1, s2) <= 0;
+        }
+
+        public static SemanticVersion Parse(string semanticVersionString)
```
Comment (nguerrera) [inline src/Microsoft.DotNet.MSBuildSdkResolver/SemanticVersion.cs]: `bool TryParse(string, out SemanticVersion)`. 

See comment about error case above.

File: src/Microsoft.DotNet.MSBuildSdkResolver/SemanticVersion.cs [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+            return this == other;
+        }
+
+        public override int GetHashCode()
+        {
+            return Major.GetHashCode() ^
+                   Minor.GetHashCode() ^
+                   Patch.GetHashCode() ^
+                   Pre.GetHashCode() ^
+                   Build.GetHashCode();
```
Comment (nguerrera) [inline src/Microsoft.DotNet.MSBuildSdkResolver/SemanticVersion.cs]: This is not a very good hash code. I'd say `(Major, Minor, Patch, Pre).GetHashCode()` but we don't want the ValueTuple dependency here.

We don't need to hash semantic versions. I'd just remove all the overrides and overloads and keep just make Compare public. Better not to have dead internal code than to have to think about things like this that don't actually matter.

File: src/Microsoft.DotNet.MSBuildSdkResolver/SemanticVersion.cs [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+
+namespace Microsoft.DotNet.MSBuildSdkResolver
+{
+    internal sealed class SemanticVersion
+    {
+        public int Major { get; private set; }
+        public int Minor { get; private set; }
+        public int Patch { get; private set; }
+        public string Pre { get; private set; }
+        public string Build { get; private set; }
```
Comment (nguerrera) [inline src/Microsoft.DotNet.MSBuildSdkResolver/SemanticVersion.cs]: These are all set in constructor so you don't need `private set`. Makes it clearer that the object is immutable.

File: src/Microsoft.DotNet.MSBuildSdkResolver/SemanticVersion.cs [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+        public SemanticVersion(int major, int minor, int patch) : this(major, minor, patch, string.Empty, string.Empty)
+        {
+        }
+
+        public SemanticVersion(int major, int minor, int patch, string pre) :
+            this(major, minor, patch, pre, string.Empty)
+        {
+        }
+
+        public SemanticVersion(int major, int minor, int patch, string pre, string build)
```
Comment (nguerrera) [inline src/Microsoft.DotNet.MSBuildSdkResolver/SemanticVersion.cs]: Just have one constructor with pre = string.Empty, build = string.Empty defaults.

File: src/Microsoft.DotNet.MSBuildSdkResolver/SemanticVersion.cs [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+                return new SemanticVersion(major, minor, patch);
+            }
+
+            if (!int.TryParse(semanticVersionString.Substring(patchStart, patchSeparator - patchStart), out patch))
+            {
+                return null;
+            }
+
+            int preStart = patchSeparator;
+            int preSeparator = semanticVersionString.IndexOf("+", preStart);
```
Comment (nguerrera) [inline src/Microsoft.DotNet.MSBuildSdkResolver/SemanticVersion.cs]: Note that fx_ver used by core-setup doesn't support this. It's fine to do it here, but we'll need to patch core-setup before we can actually ship versions like this.

File: src/Microsoft.DotNet.MSBuildSdkResolver/SemanticVersion.cs [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+
+            int minor = 0;
+            if (!int.TryParse(semanticVersionString.Substring(minorStart, minorSeparator - minorStart), out minor))
+            {
+                return null;
+            }
+
+            int patch = 0;
+            int patchStart = minorSeparator + 1;
+            int patchSeparator = semanticVersionString.FindFirstNotOf("0123456789", patchStart);
```
Comment (nguerrera) [inline src/Microsoft.DotNet.MSBuildSdkResolver/SemanticVersion.cs]: Looks like this will parse 1.0.0beta as 1.0.0-beta instead of as an error. I think this should just be IndexOfAny passing '-' or '+'.

File: src/Microsoft.DotNet.MSBuildSdkResolver/SemanticVersion.cs [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+                return string.IsNullOrEmpty(s1.Pre) ? 1 : -1;
+            }
+
+            int preCompare = string.Compare(s1.Pre, s2.Pre);
+            if (preCompare != 0)
+            {
+                return preCompare;
+            }
+
+            return string.Compare(s1.Build, s2.Build);
```
Comment (nguerrera) [inline src/Microsoft.DotNet.MSBuildSdkResolver/SemanticVersion.cs]: Semver v2 says build metadata should be ignored when considering precedence.

File: src/Microsoft.DotNet.MSBuildSdkResolver/SemanticVersion.cs [hunk 10]
```diff
@@ -0,0 +1,181 @@
...
+            {
+                return s1.Patch > s2.Patch ? 1 : -1;
+            }
+
+            if (string.IsNullOrEmpty(s1.Pre) != string.IsNullOrEmpty(s2.Pre))
+            {
+                return string.IsNullOrEmpty(s1.Pre) ? 1 : -1;
+            }
+
+            int preCompare = string.Compare(s1.Pre, s2.Pre);
```
Comment (nguerrera) [inline src/Microsoft.DotNet.MSBuildSdkResolver/SemanticVersion.cs]: CompareOrdinal

File: src/Microsoft.DotNet.MSBuildSdkResolver/SemanticVersion.cs (hunk 9 above)
Comment (nguerrera) [inline src/Microsoft.DotNet.MSBuildSdkResolver/SemanticVersion.cs]: Maybe we should simplify to v1 and only be as good as fx_ver in core-setup.

File: src/Microsoft.DotNet.MSBuildSdkResolver/SemanticVersion.cs (hunk 10 above)
Comment (nguerrera) [inline src/Microsoft.DotNet.MSBuildSdkResolver/SemanticVersion.cs]: This also doesn't handle semver v2 "1.0-alpha.10" > "1.0-alpha.9" (maybe another point to just delcare that we only support v1).

File: src/Microsoft.DotNet.MSBuildSdkResolver/SemanticVersion.cs (hunk 9 above)
Comment (nguerrera) [inline src/Microsoft.DotNet.MSBuildSdkResolver/SemanticVersion.cs]: I see now that this is in fact all ported straight from core-setup (which makes sense, if we compare differently than hostfxr, it will lead to strange behavior.) 

I think we should call it FXVersion and copy the  "// Note: This is not SemVer" from fx_ver and say that we need to match hostfxr's version comparison.

File: src/Microsoft.DotNet.MSBuildSdkResolver/SemanticVersion.cs (hunk 8 above)
Comment (nguerrera) [inline src/Microsoft.DotNet.MSBuildSdkResolver/SemanticVersion.cs]: This matches core-setup, though.

File: src/Microsoft.DotNet.MSBuildSdkResolver/MSBuildSdkResolver.cs (hunk 2 above)
Comment (livarcocc) [inline src/Microsoft.DotNet.MSBuildSdkResolver/MSBuildSdkResolver.cs]: True. Wonder what we should do if any of the versions are not formatted right though. In the resolver.

File: src/Microsoft.DotNet.MSBuildSdkResolver/MSBuildSdkResolver.cs (hunk 1 above)
Comment (livarcocc) [inline src/Microsoft.DotNet.MSBuildSdkResolver/MSBuildSdkResolver.cs]: Done.

File: src/Microsoft.DotNet.MSBuildSdkResolver/SemanticVersion.cs (hunk 5 above)
Comment (livarcocc) [inline src/Microsoft.DotNet.MSBuildSdkResolver/SemanticVersion.cs]: Done.

File: src/Microsoft.DotNet.MSBuildSdkResolver/SemanticVersion.cs (hunk 6 above)
Comment (livarcocc) [inline src/Microsoft.DotNet.MSBuildSdkResolver/SemanticVersion.cs]: Done.

File: src/Microsoft.DotNet.MSBuildSdkResolver/SemanticVersion.cs (hunk 3 above)
Comment (livarcocc) [inline src/Microsoft.DotNet.MSBuildSdkResolver/SemanticVersion.cs]: Done

File: src/Microsoft.DotNet.MSBuildSdkResolver/MSBuildSdkResolver.cs (hunk 2 above)
Comment (livarcocc) [inline src/Microsoft.DotNet.MSBuildSdkResolver/MSBuildSdkResolver.cs]: I am failing if any of the versions are not semanticly correct.

File: src/Microsoft.DotNet.MSBuildSdkResolver/SemanticVersion.cs (hunk 4 above)
Comment (livarcocc) [inline src/Microsoft.DotNet.MSBuildSdkResolver/SemanticVersion.cs]: Done.

File: src/Microsoft.DotNet.MSBuildSdkResolver/SemanticVersion.cs (hunk 9 above)
Comment (livarcocc) [inline src/Microsoft.DotNet.MSBuildSdkResolver/SemanticVersion.cs]: Done.

File: src/Microsoft.DotNet.MSBuildSdkResolver/MSBuildSdkResolver.cs [hunk 11]
```diff
@@ -53,15 +53,19 @@ public override SdkResult Resolve(SdkReference sdkReference, SdkResolverContext
                 netcoreSdkVersion = new DirectoryInfo(netcoreSdkDir).Name;;
             }
 
-            if (!string.IsNullOrEmpty(sdkReference.MinimumVersion) && 
-                SemanticVersion.Parse(netcoreSdkVersion) < SemanticVersion.Parse(sdkReference.MinimumVersion))
+            FXVersion netCoreSdkFXVersion;
+            FXVersion minimumFXVersion;
+            if (!FXVersion.TryParse(netcoreSdkVersion, out netCoreSdkFXVersion) ||
+                !FXVersion.TryParse(sdkReference.MinimumVersion, out minimumFXVersion) ||
+                netCoreSdkFXVersion < minimumFXVersion)
```
Comment (nguerrera) [inline src/Microsoft.DotNet.MSBuildSdkResolver/MSBuildSdkResolver.cs]: If netcoreSdkVersion is null here, we should consider it to be larger than any minimum. It means we got msbuildSdksDir from the environment, which is an override for forcing the Sdks to use (for testing). I don't think this override should require picking a version (though you can with a separate environment variable).

File: src/Microsoft.DotNet.MSBuildSdkResolver/FXVersion.cs [hunk 12]
```diff
@@ -0,0 +1,126 @@
+// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.DotNet.MSBuildSdkResolver
+{
+    // Note: This is not SemVer (esp., in comparing pre-release part, fx_ver_t does not
+    // compare multiple dot separated identifiers individually.) ex: 1.0.0-beta.2 vs. 1.0.0-beta.11
```
Comment (nguerrera) [inline src/Microsoft.DotNet.MSBuildSdkResolver/FXVersion.cs]: Unclear what fx_ver_t is in this context, we should point to it in core-setup via link and make note that the behavior of this class must match fx_ver_t.

File: src/Microsoft.DotNet.MSBuildSdkResolver/FXVersion.cs [hunk 13]
```diff
@@ -0,0 +1,126 @@
...
+            int preCompare = string.CompareOrdinal(s1.Pre, s2.Pre);
+            if (preCompare != 0)
+            {
+                return preCompare;
+            }
+
+            return string.CompareOrdinal(s1.Build, s2.Build);
+        }
+
+        public static bool TryParse(string FXVersionString, out FXVersion FXVersion)
```
Comment (nguerrera) [inline src/Microsoft.DotNet.MSBuildSdkResolver/FXVersion.cs]: fxVersionString (parameters should be camelCased)

File: src/Microsoft.DotNet.MSBuildSdkResolver/FXVersion.cs (hunk 12 above)
Comment (livarcocc) [inline src/Microsoft.DotNet.MSBuildSdkResolver/FXVersion.cs]: Added a link to the original fx_version file.

File: src/Microsoft.DotNet.MSBuildSdkResolver/MSBuildSdkResolver.cs [hunk 14]
```diff
@@ -53,20 +53,49 @@ public override SdkResult Resolve(SdkReference sdkReference, SdkResolverContext
...
+        private bool IsNetCoreSDKSmallerThanTheMinimumVersion(string netcoreSdkVersion, string minimumVersion)
+        {
+            FXVersion netCoreSdkFXVersion;
+            FXVersion minimumFXVersion;
+
+            return !string.IsNullOrEmpty(minimumVersion) &&
+                (!FXVersion.TryParse(netcoreSdkVersion, out netCoreSdkFXVersion) ||
+                !FXVersion.TryParse(minimumVersion, out minimumFXVersion) ||
+                FXVersion.Compare(netCoreSdkFXVersion, minimumFXVersion) == -1);
+        }
```
Comment (nguerrera) [inline src/Microsoft.DotNet.MSBuildSdkResolver/MSBuildSdkResolver.cs]: Let's split these in to ifs

File: src/Microsoft.DotNet.MSBuildSdkResolver/StringExtensions.cs [hunk 15]
```diff
@@ -0,0 +1,15 @@
+namespace Microsoft.DotNet.MSBuildSdkResolver
+{
+    internal static class StringExtensions
+    {
+        public static int FindFirstNotOf(this string s, string chars, int startIndex)
```
Comment (nguerrera) [inline src/Microsoft.DotNet.MSBuildSdkResolver/StringExtensions.cs]: I'm not a fan of extension methods on primitives, but apparently I'm alone on that on this team.

File: test/Microsoft.DotNet.MSBuildSdkResolver.Tests/Microsoft.DotNet.MSBuildSdkResolver.Tests.csproj [hunk 16]
```diff
@@ -21,6 +21,7 @@
     <PackageReference Include="Microsoft.NET.Test.Sdk" Version="15.0.0" />
     <PackageReference Include="xunit.runner.visualstudio" Version="2.2.0" />
     <PackageReference Include="xunit" Version="2.2.0" />
+    <PackageReference Include="FluentAssertions" Version="4.18.0" />
```
Comment (nguerrera) [inline test/Microsoft.DotNet.MSBuildSdkResolver.Tests/Microsoft.DotNet.MSBuildSdkResolver.Tests.csproj]: Why was this necessary? Should the version pulled in by Microsoft.DotNet.Tools.Tests.Utilities be changed instead?

**Comment by livarcocc** [comment]
@nguerrera Any idea why this would work on x64 release but fail on x86 debug?

```
GivenAnMSBuildSdkResolver.cs(73,13): error CS0012: The type 'Expression<>' is defined in an assembly that is not referenced. You must add a reference to assembly 'System.Core, Version=2.0.5.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e, Retargetable=Yes'. [D:\j\workspace\debug_windows---8a02d2bd\test\Microsoft.DotNet.MSBuildSdkResolver.Tests\Microsoft.DotNet.MSBuildSdkResolver.Tests.csproj] [D:\j\workspace\debug_windows---8a02d2bd\build.proj]
```

**Comment by nguerrera** [comment]
No idea. I'll pull it down and take a look.

File: src/Microsoft.DotNet.MSBuildSdkResolver/FXVersion.cs [hunk 17]
```diff
@@ -0,0 +1,129 @@
...
+            int preCompare = string.CompareOrdinal(s1.Pre, s2.Pre);
+            if (preCompare != 0)
+            {
+                return preCompare;
+            }
+
+            return string.CompareOrdinal(s1.Build, s2.Build);
+        }
+
+        public static bool TryParse(string fxVersionString, out FXVersion FXVersion)
```
Comment (wli3) [inline src/Microsoft.DotNet.MSBuildSdkResolver/FXVersion.cs]: nit:method too long
 int majorSeparator = fxVersionString.IndexOf(".");
 +            if (majorSeparator == -1)
 +            {
 +                return false;
 +            }
due too many times

File: src/Microsoft.DotNet.MSBuildSdkResolver/FXVersion.cs [hunk 18]
```diff
@@ -0,0 +1,129 @@
...
+
+            int minorStart = majorSeparator + 1;
+            int minorSeparator = fxVersionString.IndexOf(".", minorStart);
+            if (minorSeparator == -1)
+            {
+                return false;
+            }
+
+            int minor = 0;
+            if (!int.TryParse(fxVersionString.Substring(minorStart, minorSeparator - minorStart), out minor))
```
Comment (wli3) [inline src/Microsoft.DotNet.MSBuildSdkResolver/FXVersion.cs]: [discuss]
What do ppl think of 
https://github.com/louthy/csharp-monad#option-monad
to replace try get?

Comment (nguerrera) [inline src/Microsoft.DotNet.MSBuildSdkResolver/FXVersion.cs]: Yagni.

File: src/Microsoft.DotNet.MSBuildSdkResolver/MSBuildSdkResolver.cs [hunk 19]
```diff
@@ -53,20 +53,69 @@ public override SdkResult Resolve(SdkReference sdkReference, SdkResolverContext
...
+
+            if (FailsToParseVersions(netcoreSdkVersion, minimumVersion, out netCoreSdkFXVersion, out minimumFXVersion))
+            {
+                return true;
+            }
+
+            return FXVersion.Compare(netCoreSdkFXVersion, minimumFXVersion) == -1;
+        }
+
+        private bool FailsToParseVersions(
```
Comment (wli3) [inline src/Microsoft.DotNet.MSBuildSdkResolver/MSBuildSdkResolver.cs]: nit: Can we use fancy new tuple? There is 2 out

File: src/Microsoft.DotNet.MSBuildSdkResolver/FXVersion.cs (hunk 18 above)
Comment (nguerrera) [inline src/Microsoft.DotNet.MSBuildSdkResolver/FXVersion.cs]: TryParse is pervasive idiom in .NET. out var can make it read a little better.

File: src/Microsoft.DotNet.MSBuildSdkResolver/MSBuildSdkResolver.cs (hunk 19 above)
Comment (nguerrera) [inline src/Microsoft.DotNet.MSBuildSdkResolver/MSBuildSdkResolver.cs]: No. Can't add system.valuetuple dependency. Anyway, I think it would be clearer if inlined as mentioned elsewhere. The name is super confusing. We can get rid of extra variable declarations using out string x.

File: src/Microsoft.DotNet.MSBuildSdkResolver/StringExtensions.cs (hunk 15 above)
Comment (wli3) [inline src/Microsoft.DotNet.MSBuildSdkResolver/StringExtensions.cs]: Side note: Extension can also be use as a way of writing function chain. But I am still skeptical to write functional code in C#

Let me dig some old argument
In F# you can do things like this, (not real F#)
```F#
let itemsToReport = newLineItems
|> List.filter( fun (l) -> !l.IsPrePaid() || fun(l) -> l.IsProratedCharge())
|> Aggregate
|> RemoveReportedChargesForThisOrder
```

You can do similar stuff by extension method
```C#
var itemsToReport = newLineItems
.Where(l => !l.IsPrePaid() || l.IsProratedCharge())
.Aggregate()
.RemoveReportedChargesForThisOrder()
```

File: src/Microsoft.DotNet.MSBuildSdkResolver/FXVersion.cs (hunk 18 above)
Comment (wli3) [inline src/Microsoft.DotNet.MSBuildSdkResolver/FXVersion.cs]: Yagni apples on feature no code quality
https://martinfowler.com/bliki/Yagni.html
"Yagni is not a justification for neglecting the health of your code base. Yagni requires (and enables) malleable code."

Comment (nguerrera) [inline src/Microsoft.DotNet.MSBuildSdkResolver/FXVersion.cs]: Using the underlying base class library's idiomatic way to say "try to parse this" is not neglecting the health of your code base.

Comment (nguerrera) [inline src/Microsoft.DotNet.MSBuildSdkResolver/FXVersion.cs]: Wrapping the BCL to change syntax is YAGNI. I stand by it.

Comment (nguerrera) [inline src/Microsoft.DotNet.MSBuildSdkResolver/FXVersion.cs]: Maintaining wrappers adds to cost. This is a tiny ~100 line library.

Comment (wli3) [inline src/Microsoft.DotNet.MSBuildSdkResolver/FXVersion.cs]: Agree, in this specific place we don't need it

**Comment by wli3** [comment]
@dotnet-bot test Windows_NT x86 Debug Build

**Comment by wli3** [comment]
@dotnet-bot test Windows_NT x86 Debug Build
